File: Data_Prep/save_dna_mhe.py
# -*- coding: utf-8 -*-
"""
@author: jarre

Description:
    Takes assemblage data from .json files produced by R scripts and converts
    it to a multihot encoded array of class presence/absence per sampling event.
    The array is saved as a .csv file.
"""


from sklearn.metrics import precision_score, recall_score, confusion_matrix
import pandas as pd
import numpy as np
import json
import os

# Sample data
with open(r"C:\Carabid_Data\CV-eDNA\splits\order\image_multilab_order.json") as json_file:
    image_multilab = json.load(json_file)
with open(r"C:\Carabid_Data\CV-eDNA\splits\order\dna_multilab_order.json") as json_file:
    dna_multilab = json.load(json_file)

ground_truth = [value for value in image_multilab.values()]
predicted_labels = [value for value in dna_multilab.values()]

# Get unique labels
all_labels = sorted(list(set(label for labels in ground_truth + predicted_labels for label in labels)))

# Create label-to-index and index-to-label mappings
label_to_index = {label: index for index, label in enumerate(all_labels)}
index_to_label = {index: label for label, index in label_to_index.items()}

# Convert labels to binary arrays
def label_to_binary(label_list):
    binary_array = np.zeros(len(all_labels))
    for label in label_list:
        binary_array[label_to_index[label]] = 1
    return binary_array

ground_truth_bin = np.array([label_to_binary(labels) for labels in ground_truth])
predicted_labels_bin = np.array([label_to_binary(labels) for labels in predicted_labels])

df_array = np.column_stack((list(image_multilab.keys()), predicted_labels_bin))
df = pd.DataFrame(df_array)
df.rename(columns={0: "event"}, inplace=True)

# Save the Pandas DataFrame as a CSV file
df.to_csv('dna_mhe.csv', index=False)

File: Data_Prep/save_dna_mhe.R
# Extract ground truth and predicted labels
ground_truth = unname(image_multilab)
predicted_labels = unname(dna_multilab)

# Combine all labels and get unique labels
all_labels = sort(unique(unlist(c(ground_truth, predicted_labels))))

# Create a mapping of labels to indices
label_to_index = setNames(seq_along(all_labels), all_labels)

# Function to convert labels to binary arrays
label_to_binary = function(label_list) {
  binary_array = integer(length(all_labels))
  indices = unlist(label_to_index[label_list])
  binary_array[indices] = 1
  return(binary_array)
}

# Convert ground truth and predicted labels to binary format
ground_truth_bin = t(sapply(ground_truth, label_to_binary))
predicted_labels_bin = t(sapply(predicted_labels, label_to_binary))

# Combine event names with binary predicted labels
event_names = names(image_multilab)
df = data.frame(event = event_names, predicted_labels_bin)
